File: integrations/mcp_clients/quotabot_mcp_stdio.ts
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";

import {
  errorMessage,
  printRoutingDecision,
} from "./quotabot_mcp_common.js";

const here = dirname(fileURLToPath(import.meta.url));
const collectorDir =
  process.env.QUOTABOT_COLLECTOR_DIR ?? resolve(here, "..", "..", "collector");

const transport = new StdioClientTransport({
  command: process.env.DART ?? "dart",
  args: ["run", "bin/mcp_server.dart"],
  cwd: collectorDir,
  stderr: "inherit",
});

const client = new Client({
  name: "quotabot-mcp-stdio-example",
  version: "1.0.0",
});

try {
  await client.connect(transport);
  await printRoutingDecision(client, {
    task: process.env.QUOTABOT_TASK ?? "standard",
  });
} catch (error) {
  console.error(`quotabot MCP unavailable: ${errorMessage(error)}`);
  process.exitCode = 1;
} finally {
  await client.close().catch(() => undefined);
}

File: integrations/mcp_clients/quotabot_mcp_stdio.py
from __future__ import annotations

import asyncio
import os
import sys
from pathlib import Path

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

from quotabot_mcp_common import (
    as_pretty_json,
    require_routing_tools,
    routing_summary,
    structured_content,
)


def collector_dir() -> Path:
    configured = os.environ.get("QUOTABOT_COLLECTOR_DIR")
    if configured:
        return Path(configured).expanduser().resolve()
    return Path(__file__).resolve().parents[2] / "collector"


async def main() -> int:
    task = os.environ.get("QUOTABOT_TASK", "standard")
    server = StdioServerParameters(
        command=os.environ.get("DART", "dart"),
        args=["run", "bin/mcp_server.dart"],
        cwd=str(collector_dir()),
    )

    try:
        async with stdio_client(server) as (read_stream, write_stream):
            async with ClientSession(read_stream, write_stream) as session:
                await session.initialize()
                tools = await session.list_tools()
                require_routing_tools(tool.name for tool in tools.tools)

                suggestion = structured_content(
                    await session.call_tool("suggest_provider", {})
                )
                model = structured_content(
                    await session.call_tool("suggest_model", {"task": task})
                )
                print(as_pretty_json(routing_summary(suggestion, model)))
                return 0
    except Exception as error:
        print(f"quotabot MCP unavailable: {error}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
